Renderer immersive styling updates

nucleogenesis opened the pull request:
## Summary

After seeing that the "estimated time" was going to be in the sidebar, I went and updated the styles to match what @jtamiace [shared on 8111](https://github.com/learningequality/kolibri/issues/8111#issuecomment-914703792) for the content renderers - lots of fiddly CSS stuff.

I hacked and slashed and need to update some test suites to account for the changes that I made to some of the structure.  Will get the tests passing ASAP.

| Content Type | Before | After | After Mobile |
|-|-|-|-
| EPUB | ![before-epub](https://user-images.githubusercontent.com/6356129/134384868-90395356-944f-44af-a688-b2cf1fa99bb8.png) | ![epub-desktop-startedstatus](https://user-images.githubusercontent.com/6356129/134384888-77d5566e-ac39-47d8-8658-8d41138df9fc.png) | ![epub-mobile-just-completed-manually-with-modal](https://user-images.githubusercontent.com/6356129/134384927-49b6b382-c90c-465f-af6a-efdf729d5414.png) |
| Exercise | ![before-exercise-rtl](https://user-images.githubusercontent.com/6356129/134385252-76d32f84-33d9-4c25-bce4-058b8f22c2db.png) | ![exercise-desktop-coach-content-hints-shown](https://user-images.githubusercontent.com/6356129/134385233-85752404-cef2-412a-8094-06638616d3da.png) |![exercise-mobile-hints-shown](https://user-images.githubusercontent.com/6356129/134388813-9d2c9f4d-c3ed-41ac-a4ae-f256a0424384.png) |
| Exercise - Menu | ![exercise-context-menu-aligned-ltr](https://user-images.githubusercontent.com/6356129/134385344-f0e14580-589e-494e-af0b-37dde46bae17.png) | ![exercise-context-menu-aligned-rtl](https://user-images.githubusercontent.com/6356129/134385347-d1e0cc67-e195-4744-b346-edd121879a6d.png) | No mobile - but just noting that the context menus are aligned in both RTL and LTR :) |
| HTML5 | ![before-html5](https://user-images.githubusercontent.com/6356129/134385444-a7ffa0c9-16ba-4925-b38c-3a853bf08b84.png) | ![html5-desktop-immersif](https://user-images.githubusercontent.com/6356129/134385456-0610b6b5-da30-4d95-a75e-0ad47cb4260d.png) | ![html5-mobile-not-optimized-for-so-small-methinks](https://user-images.githubusercontent.com/6356129/134385459-762c214d-f7ed-4697-aab4-f565998b39b9.png) |
|PDF|![before-pdf](https://user-images.githubusercontent.com/6356129/134385515-cf27a511-d0ae-49f1-8387-2031b7cfdce7.png)|![pdf-desktop-so-immersive](https://user-images.githubusercontent.com/6356129/134385546-19c92f6f-7568-42a8-bc2f-c02bfdfa1294.png)|![pdf-mobile-immersive-yet-short](https://user-images.githubusercontent.com/6356129/134385548-8a4e6731-8a5b-47d3-81c9-38b1b9c56e28.png)|
| Video | ![before-video-rtl](https://user-images.githubusercontent.com/6356129/134385572-bc87997a-1ffb-4fce-a47f-f38ae10ba4c9.png) | ![video-desktop-just-completed-manually-with-modal](https://user-images.githubusercontent.com/6356129/134385576-4e42011c-e740-4286-b222-c2994382c530.png) | ![video-mobile-startedstatus](https://user-images.githubusercontent.com/6356129/134385577-deac1d71-4d12-457d-8629-7b9c7cceddd5.png)|



## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

[The image in this comment](https://github.com/learningequality/kolibri/issues/8111#issuecomment-914703792)

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

This touches a lot of layout spacing things - so please try testing at various sizes on multiple devices if you can.

- General layout and spacing, alignment of top/bottom bars with content.
- Icon spacing & alignment
- Tooltip alignment & visibility
- Hints on exercises
- Test a bunch of Perseus exercises if you can - especially content NOT in the QA channel (the more diverse a selection of content the better please).

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
add padding to video player per specs

nucleogenesis committed:
Fix wrong property name in TranscriptMenuItem for string

nucleogenesis committed:
Show progress icon in learn immersive activity bar

nucleogenesis committed:
Misc cleanup

- reinstate `IsTopic` computed property
- remove duplicate KCircularLoader
- instantiate content and channel from Vuex

nucleogenesis committed:
BottomAppBar CSS tweaks for immersive layout

nucleogenesis committed:
EpubRenderer styles update for immersive

nucleogenesis committed:
remove lessonmasterybar from assessmentwrapper

nucleogenesis committed:
Top bars spacing, styling, hint layout

nucleogenesis committed:
PDF Renderer immersive styles

nucleogenesis committed:
LessonMasteryBar, hints, and CSS for immersive on Perseus Renderer

nucleogenesis committed:
Remove the Hint component

nucleogenesis committed:
CoachContentLabel tooltip on bottom - not pos rel

nucleogenesis committed:
UI updates across exercise renderer components & top bar for all

nucleogenesis committed:
vuex-ify learning-activity-bar test suite

nucleogenesis committed:
cleaning up a few messes

MisRob commented on kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:
```diff
@@ -170,6 +183,12 @@
       };
     },
     computed: {
+      ...mapState({
```
I'd suggest not adding dependencies to Vuex in smaller components that don't need it like this learning activity bar. I don't see a reason why not to use  `props` here that are made for that purpose, especially when the amount of information you're adding here is not huge nor complex. This will also simplify the test suite that will otherwise have tendencies to grow complicated even for very simple tests (you won't need `makeStore` and you won't need to know store structure to test a simple UI piece)

MisRob replied:
Another good reason to avoid is that we're starting using Composition API and it'd be good to keep smaller building block components flexible enough to be able to work well no matter if their more complex parent components are built with Vuex or with Composition API.

nucleogenesis replied:
Done and pushed - thanks for catching that Misha!

nucleogenesis committed:
Push Vuex stuff to parent, pass to LearningActivityBar via props

rtibbles commented on kolibri/core/assets/src/core-app/apiSpec.js:
```diff
@@ -101,6 +101,9 @@ import InteractionList from '../views/InteractionList';
 import ExamReport from '../views/ExamReport';
 import TextTruncator from '../views/TextTruncator';
 import TimeDuration from '../views/TimeDuration';
+// TODO Migrate SuggestedTime into ../views/SuggestedTime. This is not to be done
```
We can do this! As long as the component name does not change, the string ids will be constant - and messages are extracted from any CSV file, not just the plugin specific one.

rtibbles commented on kolibri/plugins/perseus_viewer/assets/src/views/PerseusRendererIndex.vue:
```diff
@@ -1,22 +1,16 @@
+    class="bibliotron-exercise perseus-root"
+    :class="{ 'perseus-mobile': isMobile }"
+  >
+    <LessonMasteryBar
```
I _think_ this is the right place for this - although if it's something we would expose in, say, the QTI renderer too, then we might need to move it out of here in the future.

nucleogenesis replied:
If we want it in the QTI renderer we can move it to Core and update the imports for sure if we keep that part of the design consistent between Perseus & QTI.

rtibbles commented on kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue:
```diff
@@ -407,15 +407,14 @@
-    // This ensures that showing vs hiding the controls
-    // will not cover visible content below (ie, author name)
-    margin-bottom: $controls-height;
+    height: calc(100vh - #{$top-bar-height} - #{$controls-height});
```
Woot!

rtibbles commented on kolibri/plugins/learn/assets/src/views/classes/Hint.vue:
```diff
@@ -1,59 +0,0 @@
-<template>
```
I wonder if we could just leave this as is, and instead remove the hint button inside the Perseus Renderer, and then just hook things up via props and events via the `LessonMasteryBar`?

That would leave us open for reusing this in say, the QTI renderer if needed.

nucleogenesis replied:
I'd rather not keep the cruft around in case it's maybe useful. The component isn't usable as-is and is only used in `LessonMasteryBar` so I don't see the need for a reusable component - particularly if the QTI Renderer ends up finding `LessonMasterBar` useful in its design, in which case the hint code will be reused as-is.

rtibbles reviewed (commented):
Some suggestions for tweaks, but nothing blocking.

jtamiace commented:
I'm putting my comments in this deck because I don't want to overload this PR with screenshots. Hope this doesn't make things difficult to track @nucleogenesis - let me know if this isn't a helpful way to document requested changes. It's all looking great so far and feeling really nice to interact with 👏

https://docs.google.com/presentation/d/1eDQ4NFgD1suKOnAALWbV4L_3KX_VSQlHyDKFZRYD14M/edit#slide=id.p

nucleogenesis commented:
@jtamiace thank you for that! I went through and posted a few comments back to you with a few questions. Otherwise, I'll get started on updates.

nucleogenesis committed:
Move SuggestedTime into core api

nucleogenesis committed:
Renderers at full-height across the board, misc ui fixes

EPUB
- 1px tokens.fineLine border
- full height on all screens (minus padding per specs)

PDF
- Automatically scales the page back the larger the screen per our responsive screen mixin sizes. This is done by applying a multiplier to the denominator of the math establishing the initial width. We divide an HTML elements width by the width of the first page - and now we make the width higher by 1x on small, 1.25x on large and 1.125x for in between, effectively zooming out of the page when it loads so that the user sees more page content at a normal size (or closer to it) than before.

EXERCISES
- Removed gold star & checkmark from LessonMasteryBar
- Applied 1px tokens.fineLine to LearningActivityBar in place of the normal drop shadow - only on exercises. This gives the two top bar appearance a more unified look as if there are two sections of one overall top bar.
- Max width of 1200px applied to the content window, left aligned (right in RTL).

HTML5
- Takes up full height and width of the available screen, permits scrolling for tall content (ie, Flexbook)  so that it looks and feels natural.
- This was done by removing all of the watchers/activity around Hashi watching for resizing.

radinamatic commented:
The `View lesson resources` icon is not going to be permanent, correct?

I have no users, classes, lessons or quizzes on this facility (fresh one, viewing the resource from the Library as a super admin), so makes no sense to have the icon `View lesson resources` during the free content exploration... 🤔 

![LEDev2104 (start)  Running  - Oracle VM VirtualBox_046](https://user-images.githubusercontent.com/1457929/135376038-d7d1517e-654c-489b-a1a1-0a9b3ba60bc9.png)

radinamatic commented:
Video transcript is not loading:

![LEDev2104 (start)  Running  - Oracle VM VirtualBox_048](https://user-images.githubusercontent.com/1457929/135376509-69347a09-39ac-422b-9c42-cf251e75c504.png)

radinamatic commented:
Not strictly content renderer, but not sure of these missing padding on the Library page on tablet screen is related to the CSS changes...? (both FF & Chrome on Linux)

![LEDev2104 (start)  Running  - Oracle VM VirtualBox_049](https://user-images.githubusercontent.com/1457929/135377246-dd63c508-055f-4586-8452-daf77aa0b217.png)


Idem for this jumbled sections while browsing channel on tablet screen 😕 

![LEDev2104 (start)  Running  - Oracle VM VirtualBox_050](https://user-images.githubusercontent.com/1457929/135379222-920aafec-da29-4b3f-a2b7-c72aea92d839.png)

radinamatic commented:
Exercises generally look good, some glitches on smaller mobile resolutions:

1 | 2
--- | ---
![exercise2](https://user-images.githubusercontent.com/1457929/135384801-289526c5-6f4c-4125-b8c8-bedab0164234.gif) | ![exercise](https://user-images.githubusercontent.com/1457929/135384804-1f87caa1-0b80-4376-9d05-244efd22e538.gif)

radinamatic commented:
Btw, I was not able to access the above server running on my dev VM from any other device in my local network 🤔 

`Uncaught ReferenceError: kolibriCoreAppGlobal is not defined`

![Kolibri — Mozilla Firefox_003](https://user-images.githubusercontent.com/1457929/135385181-9d78a0ed-29aa-4d9c-8fa7-8b5ba74203f5.png)

nucleogenesis commented:
> Not strictly content renderer, but not sure of these missing padding on the Library page on tablet screen is related to the CSS changes...? (both FF & Chrome on Linux)

This section isn't in scope for the PR - I think that @marcellamaki or @MisRob are working on those issues elsewhere (if not then they're upcoming).

radinamatic commented:
One last thing regarding the `Save to bookmark` button, @nucleogenesis, that I did not mention because I was not sure if it happens because the feature is not fully implemented: resource is correctly saved under Bookmarked when the button is pressed, but that _pressed_ state seems to stick even when the user opens another resource, so when the user hovers over the button, the label now says `Remove from bookmarks`, even though that second resource was never saved before.

marcellamaki commented:
@radinamatic - good eye on both the library page layout and the "save to bookmarks" button! I'm including fixes for those in two upcoming PRs and will definitely 👋 at you for testing them then to be sure they are good to go

rtibbles commented:
This bookmark stickiness was probably because of an incorrect assumption I had made about the internal state of the top bar being torn down between rendering content items, so the simplest fix would be to clear the bookmark data when the content being displayed is updated.

pcenov commented:
@radinamatic happy to report that I tested a number of resources in IE11 and also with RTL languages and did not stumble on any new issues other than the ones already reported. :) Noting that it's currently not possible to view any of the assigned resources when being logged in as a learner, so I was testing only at http://127.0.0.1:8080/en/learn/#/library

nucleogenesis committed:
incorrect common string fn name used, fixes vieo transcripts

nucleogenesis committed:
Misc fixes in renderers in response to feedback (details in commit msg)

- BottomAppBar remove bottom padding so that feedback text can appear under the underscores (add overflow rules accordingly)

- Remove errant space at bottom of PDF renderer

- Align all top bars on left/right side properly after fixing issue with long hint text in bg-bg language where the screen would become wider, adding unwanted scroll bars.

- Misc Perseus styling fixes that digs into Perseus styling internals a bit to center things (padding on tables) and to ensure that drag-and-drop bits stay within the given bounds and don't go off screen.

- Code which kept tooltip on screen broke other tooltips... KTooltip needs some work desperately.

- Flipped `isLessonContext` to false to see the string other than "view lesson resources" but did not wire it up to anything

nucleogenesis commented:
Everything mentioned here (and to me on Slack by @jtamiace ) are resolved in my last commit (or I hope they are anyway).

Feedback points needed:

- isLessonContext - I didn't dig into "how do we know this" yet but have guesses. Not sure this is in scope here so we can follow up on it.
- I manually mess w/ some styling on the HTML tag in PDF renderer and it's hacky but works - so if you could just try to break a PDF that'd be awesome.


Separately, my "tool  tip stays on the page" fix broke them elsewhere so I removed it :( KTooltip needs some TLC in general

nucleogenesis commented:
Also, hints now will look decent in any language:

![image](https://user-images.githubusercontent.com/6356129/135928148-e51539fe-019e-4379-a452-0c1fe9cd518a.png)

![image](https://user-images.githubusercontent.com/6356129/135928208-2257d33f-7df4-438b-bb1a-75e120f23ead.png)

nucleogenesis committed:
fix hints reflowing on smaller screens

radinamatic commented:
@pcenov Could you give one more run with the latest asset from this PR to check responsiveness on the PDF and Perseus exercises, before we merge? Thank you!

pcenov commented:
Hi @radinamatic, here are the only 2 issues I was able to find regarding the PDFs and Perseus exercises
1. The menu with more options appears under the full screen header of the PDF files and the tooltip 'More options' should not be displayed in mobile as it remains visible as long as the menu is displayed.
2. The text of the tooltip message which is shown after tapping the info icon for the hint is never displayed in full size regardless of the resolution.

| PDF | Exercise
| --- | ---
|  ![2021-10-05_13-42-17](https://user-images.githubusercontent.com/79847249/136011211-b361b7a3-7059-41d2-8176-278dcfe375e9.png) | ![2021-10-05_13-45-50](https://user-images.githubusercontent.com/79847249/136011378-669c2e05-1c23-4afa-b51d-d96b85337d6c.png)

nucleogenesis commented:
@pcenov thank you! Great catch on the PDF bar overlaying the menu!

I've fixed that problem and have finally found a way to hack that tooltip into place without messing anything else up.

https://user-images.githubusercontent.com/6356129/136110339-ec66efa5-25ae-4c0d-8f2c-2df3b7a7af2e.mp4

As for the tooltip sticking on mobile and when the menu is opened - I'm not sure the solution here and I don't think fixing it is in scope for this PR. The KTooltip component needs to be able to show on mobile because it is used in our info icon which can be clicked to show a tooltip elsewhere.

I won't say this is the best solution - but perhaps we could add a prop to KTooltip which says "don'tShowOnMobile" or something so that we can give devs that option where we will explicitly not show the tooltip in a mobile (or touch, probably) context.

That could be followed up on in Kolibri Design System the best, I think.

nucleogenesis committed:
tooltip stays on screen without breaking anything in perseus, pdf heading reduced z-index

pcenov commented:
@nucleogenesis thank you as well - I've retested and verified that both issues are now fixed. Awesome!

radinamatic reviewed (approved):
Manual testing successful and passing, great job! 👍🏽 💯
